subsys/net/lib/sockets/sockets_can.c

```diff
@@ -398,13 +541,46 @@ static int can_sock_setsockopt_vmeth(void *obj, int level, int optname,
 		}
 
 		if (optname == CAN_RAW_FILTER) {
-			struct zcan_filter zfilter;
-
-			can_copy_filter_to_zfilter((struct can_filter *)optval,
-						   &zfilter);
-
-			return api->setsockopt(dev, obj, level, optname,
-					       &zfilter, sizeof(zfilter));
+			int count, i;
+
+			if (optlen % sizeof(struct can_filter) != 0) {
+				errno = EINVAL;
+				return -1;
+			}
+
+			count = optlen / sizeof(struct can_filter);
+
+			ret = can_register_filters(iface, obj,
+						   optval, count);
+			if (ret < 0) {
+				errno = -ret;
+				return -1;
+			}
+
+			for (i = 0; i < count; i++) {
+				struct zcan_filter zfilter;
+
+				can_copy_filter_to_zfilter(
+					&((struct can_filter *)optval)[i],
+					&zfilter);
+
+				ret = api->setsockopt(dev, obj, level, optname,
```

Comment:
You should check here if someone else already attached the same filter. If yes, there is no need to attach it again.